core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/optimizer/QueryModelNormalizerOptimizer.java

```diff
@@ -231,13 +231,40 @@ private boolean isWellDesigned(LeftJoin leftJoin) {
 			leftJoin.getCondition().visit(optionalVarCollector);
 		}
 
-		Set<String> problemVars = optionalVarCollector.getVarNames();
-		problemVars.removeAll(leftJoin.getLeftArg().getBindingNames());
+		Set<String> leftBindingNames = leftJoin.getLeftArg().getBindingNames();
+		Set<String> problemVars = retainAll(optionalVarCollector.getVarNames(), leftBindingNames);
 
 		if (problemVars.isEmpty()) {
 			return true;
 		}
 
+		return checkAgainstParent(leftJoin, problemVars);
+	}
+
+	private Set<String> retainAll(Set<String> problemVars, Set<String> leftBindingNames) {
+		if (!leftBindingNames.isEmpty() && !problemVars.isEmpty()) {
+			if (leftBindingNames.size() > problemVars.size()) {
+				for (String problemVar : problemVars) {
+					if (leftBindingNames.contains(problemVar)) {
+						HashSet<String> ret = new HashSet<>(problemVars);
+						ret.removeAll(leftBindingNames);
+						return ret;
+					}
+				}
+			} else {
+				for (String leftBindingName : leftBindingNames) {
+					if (problemVars.contains(leftBindingName)) {
+						HashSet<String> ret = new HashSet<>(problemVars);
+						ret.removeAll(leftBindingNames);
+						return ret;
+					}
+				}
+			}
+		}
+		return problemVars;
+	}
```

Comment:
This is the fix for the reported bug. I've also taken the liberty to check other places where we are assuming that a returned set is modifiable (which I feel is a bad assumption).

This code is covered by the tests in the SPARQL compliance module.